salt/tests/unit/modules/files/test_metalk8s_kubernetes_utils.yaml

```diff
@@ -0,0 +1,105 @@
+  - kwargs: {"context": "my-kwargs-mycontext"}
+    api_server_pillar: {"kubeconfig": "my-kwargs-kubeconfig.conf"}
+    result: ['my-kwargs-kubeconfig.conf', 'my-kwargs-mycontext']
+  # 6. get kubeconfig partly from salt config options and context result is None
+  - config_options: {"kubernetes.kubeconfig": "my-config-option-kubeconfig.conf"}
+    result: ['my-config-option-kubeconfig.conf', null]
+  # 7. get kubeconfig partly from api-server pillar and context result is None
+  - api_server_pillar: {"kubeconfig": "my-pillar-kubeconfig.conf"}
+    result: ['my-pillar-kubeconfig.conf', null]
+  # 8. get kubeconfig partly from kwargs and context result is None
+  - kwargs: {"kubeconfig": "my-kwargs-kubeconfig.conf"}
+    result: ['my-kwargs-kubeconfig.conf', null]
+  # 9. test that kwargs overrides pillar and salt config option
+  - kwargs: {"kubeconfig": "my-kwargs-kubeconfig.conf", "context": "my-kwargs-mycontext"}
+    api_server_pillar: {
+      "kubeconfig": "my-pillar-kubeconfig.conf",
+      "context": "my-pillar-context"
+    }
+    config_options: {"kubernetes.kubeconfig": "my-config-option-kubeconfig.conf"}
+    result: ['my-kwargs-kubeconfig.conf', 'my-kwargs-mycontext']
+  # 10. test that kubeconfig is defined in the lowest level and context is being overridden
+  - api_server_pillar: {"context": "my-pillar-context"}
+    config_options: {
+      "kubernetes.kubeconfig": "my-config-option-kubeconfig.conf",
+      "kubernetes.context": "my-config-option-mycontext"
+    }
+    result: ['my-config-option-kubeconfig.conf', 'my-pillar-context']
+
+read_and_render_yaml_file:
+  # 1. read and render a valid yaml file
+  - source: |-
+      apiVersion: v1
+      kind: Pod
+      metadata:
+        name: kube-apiserver
+    result:
+      apiVersion: v1
+      kind: Pod
+      metadata:
+        name: "kube-apiserver"
+  #. 2. Raise error when no source file is specified
+  - source: null
+    raises: True
+    result: "Source file 'my-source-file' not found"
+  # 3. Raise error when a valid template is given but salt is unable to render the file
+  - source: |-
+      apiVersion: v1
+      kind: Pod
+      metadata:
+        name: kube-apiserver
+    template: 'jinja'
+    raises: True
+    result: 'Failed to render file path with error'
+  # 4. Raise error when the salt template is invalid or unrecognized
+  - source: |-
+      apiVersion: v1
+      kind: Pod
+      metadata:
+        name: kube-apiserver
+    template: 'my_jinja'
+    raises: True
+    result: 'Unknown template specified'
+  # 5. Test jinja templating with context values
+  - source: |-
+      apiVersion: v1
+      kind: Pod
+      metadata:
+        name: {{ name }}
+    template: 'jinja'
+    context:
+      name: kube-apiserver1
+      opts:
+        cachedir: "/path/to/cache/dir"
+        extension_modules: '/path/to/dummy/extension/module'
+        file_client: 'local'
+        file_ignore_glob: null
+        fileserver_backend: null
```

Comment:
No, it's too strange to give them from context always patch opts in python please

```
patch.dict(metalk8s_kubernetes_utils.__opts__, opts_dict)
```
```